packages/repository/src/__tests__/unit/repositories/define-repository-class.unit.ts

```diff
@@ -0,0 +1,166 @@
+// Copyright IBM Corp. 2020. All Rights Reserved.
+// Node module: @loopback/repository
+// This file is licensed under the MIT License.
+// License text available at https://opensource.org/licenses/MIT
+
+import {Constructor} from '@loopback/context';
+import {expect} from '@loopback/testlab';
+import {
+  AndClause,
+  AnyObject,
+  Condition,
+  Count,
+  CrudRepository,
+  DataObject,
+  DefaultCrudRepository,
+  DefaultKeyValueRepository,
+  defineCrudRepositoryClass,
+  defineEntityCrudRepositoryClass,
+  defineKeyValueRepositoryClass,
+  Entity,
+  Filter,
+  juggler,
+  model,
+  Model,
+  OrClause,
+  property,
+} from '../../..';
+
+describe('RepositoryClass builder', () => {
+  describe('defineCrudRepositoryClass', () => {
+    it('should generate CRUD repository class', async () => {
+      const AddressRepository = defineCrudRepositoryClass(
+        Address,
+        DummyCrudRepository,
+      );
```

Comment:
I find this usage problematic, here is what VSCode shows to me in a type hint:

```ts
const AddressRepository: CrudRepositoryClass<Address, DummyCrudRepository<any>>
```

Notice that `DummyCrudRepository` is using type `any` instead of `Address`. This is going to disable type checks in code using address repository instances.

```ts
const repo = new AddressRepository(db);
const address = await repo.create({/*...*/});
// boom, address has type `any` and the compiler don't warn us at all
```

Reply:
Unfortunately, I was not able to find a reliable way how to force TypeScript to reject such usage and always force the user to provide all generic arguments. It seems to me that TypeScript thinks it's able to infer those generic args, but the inference does not work as intended :(